TickSpec/TickSpec.fs

```diff
@@ -98,69 +111,103 @@ type StepDefinitions (givens,whens,thens,events,valueParsers) =
         StepDefinitions(assembly.GetTypes())
     /// Constructs instance by reflecting against specified types
     new (types:Type[]) =
-        let getScope attributes =
-            attributes
-            |> Seq.cast
-            |> Seq.fold (fun (tags,features,scenarios) (x:StepScopeAttribute) ->
-                x.Tag::tags, x.Feature::features, x.Scenario::scenarios
-            ) ([],[],[])
         let methods =
             types
-            |> Seq.collect (fun t ->
-                let attributes = t.GetCustomAttributes(typeof<StepScopeAttribute>,true)
-                let tags, features, scenarios = getScope attributes
-                t.GetMethods()
-                |> Seq.map (fun m ->
-                    let attributes = m.GetCustomAttributes(typeof<StepScopeAttribute>,true)
-                    let tags', features', scenarios' = getScope attributes
-                    tags@tags' |> List.filter (not << String.IsNullOrEmpty),
-                        features@features' |> List.filter (not << String.IsNullOrEmpty),
-                            scenarios@scenarios' |> List.filter (not << String.IsNullOrEmpty),
-                                m
+            |> Seq.collect (fun t -> t.GetMethods())
+        StepDefinitions(methods)
+    internal new (methods:MethodInfo seq) =
+        let categorizedMethods =
+            let getScope attributes =
+                attributes
+                |> Seq.cast
+                |> Seq.fold (fun (tags,features,scenarios) (x:StepScopeAttribute) ->
+                    x.Tag::tags, x.Feature::features, x.Scenario::scenarios
+                ) ([],[],[])
+            
+            let attributeMap = Dictionary<Type, (MethodWithScope * obj) list>()
+            let parentScope = Dictionary<Type, MethodScope>()
+            let methodScope = Dictionary<MethodInfo, MethodScope>()
+
+            let attributes = [|
+                typeof<GivenAttribute>; typeof<WhenAttribute>; typeof<ThenAttribute>
+                typeof<BeforeScenarioAttribute>; typeof<AfterScenarioAttribute>
+                typeof<BeforeStepAttribute>; typeof<AfterStepAttribute>
+                typeof<ParserAttribute>
+            |]
+
+            // Initialize the attribute map
+            attributes |> Array.iter (fun attrType -> attributeMap.[attrType] <- List.empty)
+
+            // Iterate through all methods
+            methods |> Seq.iter (fun mi ->
```

Comment:
for FTW
```
for mi in methods do
    for attr in mi.GetCustomAttributes true do
       match .... with
       | None -> ()
       | Some attrType ->
       
       <action>
```     
i.e. unindent for the normal case - this is considered idiomatic formatting and makes this far more legible without wide screen!

Reply:
Cannot use for-loop in constructor due to weird limitation of the older F# that is used in .NET Framework (see https://github.com/dotnet/fsharp/issues/13533). Had it using for loops originally :)

Have unindented the code. Good point, haven't been aware of this possibility.